### ISSUE TITLE
Support running each test file in a separate processes

### ISSUE THREAD

**Issue by ZyX-I** [Olivine-Labs/busted#278]
Reasons:

1. Existence of lua global state may easily lead to tests working “by accident”: just because previously run test has modified environment in some way next test works. Ref neovim/neovim#1144.
2. On a multi-core machine tests running in parallel may benefit from multiple cores since lua is single-threaded. This is actually a separate issue: running in separate processes does not imply running these processes in parallel. But it is better to think about parallel run before adding this feature, not after; especially given that running processes in a sequence will result in lots of overhead on process spawn and environment initialization (in case of neovim it is not only lua initialization), while parallel run may actually speed this up on most of the developer machines.

**Comment by DorianGray** [comment]
This will be solved with threads along with the new busted context features that prevent global leakage from one test/library to another.

**Comment by justinmk** [comment]
Great news!

**Comment by ZyX-I** [comment]
Threads and new context features are completely useless for preventing sharing global state which lives in a *C* library imported by luajit ffi. In neovim/neovim#5119 I am starting to experience problems with this: running one file is OK, running the whole test suite crashes in an unrelated position because of the luajit GC. I also cannot determine whether suite succeeded because there is some state left from previous test or because test is correct.

**Comment by DorianGray** [comment]
Yeah, there's nothing we can do to fix that. It has to do with the way luajit ffi works.

**Comment by ZyX-I** [comment]
@DorianGray I would not say so, see https://github.com/ZyX-I/neovim/commit/203c2c3e0ec992962b734ca228819ceebc55eb4f. Shortly: it just forks in each `it` and uses pipes to transform errors. This, of course, does not care about hooks like after_each which are still run in the master process. Does not distinguish between “error” and “failure”. Reports errors back as a string, and if traceback is not enough or formatter does something fancy with error it will not be able to do this. Etc.

Most of the problems may be fixed by those who are more familiar with busted and lua, but running tests in a separate process is possible.

**Comment by ZyX-I** [comment]
“Of course” is “of course” here only because I do not want to modify busted, not because I see some technical problems with running hooks inside a child process.

**Comment by DorianGray** [comment]
I shouldn't have said nothing. I wrote that solution off due to the following:

 I see a lot of performance issues there and also issues with things like process ids changing(specifically, where one it() generates some data dependent on some process level resource(like id, specifically UUIDs would have this issue), puts it in a describe level variable, and then you compare in another it().

This is why I wanted to do process per file, to avoid these issues and to increase performance through concurrency, but it doesn't solve your immediate issue with luajit. The real solution I think is to fix luajit ffi so it can reload libraries... which is also non-trivial to do.

**Comment by ZyX-I** [comment]
I did not see any real performance drop for some reason (busted reports time changed from ~0.27 to ~0.38 AFAIR, but zsh `time` does not report anything like this). I can’t suggest anything for tests written with a dependency on the process they are run in, but this should be uncommon and in any case I would expect running tests in a separate processes optional (the fact that I do not see any performance issues does not mean there will be no, also some people may consider tests depending on the execution order due to shared state being not something that needs fixing in their test suite). I have never heard of reloading libraries, guess because this is exactly a thing which is too hard to implement and too easy to work around by simply starting another process.

Also reloading a library may be not a solution: most problems I see are due to luajit GC, separate processes mean that I do not need to care about memory leaks (if I will constantly leak memory throughout the whole test suite this may become a problem, but not memory leaks in individual tests) and investigate why GC makes my suite crash (who knows whose bug is this, I better smash problem with `fork` and solve resulting issues because it also solves problem with state then spend time determining how to fix just my tests).
